### PR TITLE
feat(freeze): major changes to simplify api, freeze objects

### PR DESCRIPTION
**Pull request by ksafranski**
Major changes to API based on #2:

- Removes strategy pattern, treating all fixtures the same using `Proxy` to override `set`'s and `get`'s
- Adds `create` method which will (explicitly) produce a mutated version of an existing object
- Simplifies getting of objects by removing `create` and instead making the native objects properties of the `fixd` parent

Needed before merge:

- [ ] IRL testing (@levithomason)
- [x] Reworking of documentation
- [x] Additional integration tests

### CODE REVIEW

File: fixd.js [hunk 1]
```diff
@@ -1,14 +1,30 @@
...
+  /**
+   * Add a fixture to the fixd library and freezes if not in namespace
+   * @param {String} name A string name to represent the fixture
+   * @param {*} value The fixture value
+   */
+  add: (name, val) => {
+    if (Object.keys(fixd).indexOf(name) >= 0) {
+      throw new Error(`Cannot add fixture to already reserved namespace '${name}'.`)
+    }
+    fixd[name] = Object.freeze(val)
```
Comment (levithomason) [inline fixd.js]: Note this will only work with objects and arrays.  If we want to support primitives such as strings and numbers, we'll have to have some kind of object wrapper cache internally.

FWIW, I'm fine with an object/array for the simplicity and initial release.  The user can just format their string as an object if needed.

File: fixd.js [hunk 2]
```diff
@@ -1,14 +1,30 @@
...
+  },
+  /**
+   * Allows mutation of already existing fixture by providing name of fixture and
+   * modifier method
+   * @param {String} name The namespace of the fixture to mutate
+   * @param {Function} modifier A function to modify the fixture
+   */
+  mutate: (name, modifier) => {
+    if (!modifier || typeof modifier !== 'function') {
```
Comment (levithomason) [inline fixd.js]: "not a function" covers `!modifier`

File: fixd.js (hunk 1 above)
Comment (ksafranski) [reply fixd.js]: It won't freeze it, but it won't _not_ work

File: src/index.js [hunk 3]
```diff
@@ -1,81 +0,0 @@
...
-   * @param {...Array} args Any plugin-specific arguments
-   * @returns {*}
-   */
-  create: (name, ...args) => {
-    if (!fixd.fixtures[name]) throw new Error(`Fixture ${name} was not found, please add before creating a new instance`)
-    return fixd.fixtures[name].create(fixd.fixtures[name].data, ...args)
-  }
-}
-
-module.exports = fixd
```
Comment (levithomason) [inline src/index.js:81]: Woot! 🦉

File: fixd.js (hunk 1 above)
Comment (ksafranski) [reply fixd.js]:
```
> const foo = Object.freeze('bar')
undefined
> console.log(foo)
bar
```

File: test/fixd.spec.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+chai.use(dirtyChai)
+
+// Source
+const fixd = require('../fixd')
+
+describe('fixd', () => {
+  describe('add', () => {
+    it('throws an error if attempting to assign to already reserved namespace', () => {
```
Comment (levithomason) [inline test/fixd.spec.js]: Note that frozen objects only throw in `'use strict'` pragma environments.  This is the default for a Node module as well, however, it won't _always_ throw, such as when used in a browser without `'use strict'` set.

File: fixd.js (hunk 1 above)
Comment (ksafranski) [reply fixd.js]: oh shit, yeah, I know what you mean... I'm ok with objects and arrays, realistically a primitive assigned to a `const` will throw with `TypeError: Assignment to constant variable` anyways...

File: fixd.js [hunk 5]
```diff
@@ -1,14 +1,30 @@
...
+    }
+    fixd[name] = Object.freeze(val)
+  },
+  /**
+   * Allows mutation of already existing fixture by providing name of fixture and
+   * modifier method
+   * @param {String} name The namespace of the fixture to mutate
+   * @param {Function} modifier A function to modify the fixture
+   */
+  mutate: (name, modifier) => {
```
Comment (dvdzkwsk) [inline fixd.js]: Talked with @levithomason about this, I hate to bikeshed but I think given the API structure this method name is misleading. Because values are added to the store w/ a method call, and accessed with a getter, it needs to be even _more_ clear that `mutate` acts as a getter and not as something that acts on fixd's internal store. The language of `mutate` implies that it will modify what's in the global store.

```js
fixd.add('thing', {})          // adds 'thing' to fixd's store
fixd.thing                     // returns 'thing'
fixd.mutate('thing', /* .. */) // returns 'thing' w/ modifications
```

I think something that is more clear about it being a return value would benefit API sanity. One idea would be `fixed.withMutations()` since this implies that it acts more like a getter.

File: fixd.js [hunk 6]
```diff
@@ -1,14 +1,30 @@
...
+    if (!modifier || typeof modifier !== 'function') {
+      throw new Error('Modifier must be a function')
+    }
+    return Object.freeze(modifier(merge({}, fixd[name])))
```
Comment (dvdzkwsk) [inline fixd.js]: `Object.freeze` is not recursive, so we'll need to use a helper to make sure the result is deeply frozen.

File: fixd.js (hunk 1 above)
Comment (dvdzkwsk) [inline fixd.js]: `Object.freeze` is not recursive, so we'll need to use a helper to make sure the result is deeply frozen.

**Comment by levithomason** [comment]
🎉  image

The intent and high level look good to me, needs recursive freezing.  Pre agreeing to David's coming comments.

Once those are addressed, this can be tested in Pipehub more extensively.  I'll just install from this branch.

File: fixd.js [hunk 7]
```diff
@@ -1,14 +1,30 @@
...
-} catch (e) {
-  throw new Error('Error loading built-in plugins', e.message)
+const fixd = {
+  /**
+   * Add a fixture to the fixd library and freezes if not in namespace
+   * @param {String} name A string name to represent the fixture
+   * @param {*} value The fixture value
+   */
+  add: (name, val) => {
+    if (Object.keys(fixd).indexOf(name) >= 0) {
```
Comment (dvdzkwsk) [inline fixd.js]: Can save yourself the iteration for an operation that is a simple map lookup.

```js
if (fixd[name])   // idea #1
if (name in fixd) // idea #2
```

File: fixd.js (hunk 5 above)
Comment (levithomason) [reply fixd.js]: Agree `add` adds a fixture to fixd, `mutate` seemingly should mutate a fixture in fixd.  I'm good with anything clear and agreeable to others.  More ideas:

```js
fixd.create()
fixd.copy()
fixd.callbackWithTheOriginalButLetMeChangeIt()
```

File: fixd.js (hunk 7 above)
Comment (ksafranski) [reply fixd.js]: duh

File: fixd.js (hunk 5 above)
Comment (ksafranski) [reply fixd.js]: I like `withMutations` but seems a bit verbose, perhaps use `create`, then make the modifier optional, so the API is consistent but flexible. To visualize:

```javascript
fixd.add('foo', { fizz: 'buzz' })

fixd.foo // -> { fizz: 'buzz' }
fixd.create('foo') // -> { fizz: 'buzz' }
fixd.create('foo', () => { fizz: 'baz' }) // -> { fizz: 'baz' }
```

File: package.json [hunk 8]
```diff
@@ -52,6 +52,7 @@
     ]
   },
   "dependencies": {
+    "deep-freeze": "^0.0.1",
```
Comment (levithomason) [inline package.json]: There are a host of unresolved [issues](https://github.com/substack/deep-freeze/issues) with no responses on this module.  Given it is a single [12 LOC](https://github.com/substack/deep-freeze/blob/master/index.js) function, I'm wondering if we ought to just copy source into `fixd` so we can make patch updates.

File: fixd.js (hunk 5 above)
Comment (levithomason) [reply fixd.js]: Sounds good.

> 🥘  <kbd>4</kbd> 💭 
Variadic arguments remove the ability for function composure.  A required callback makes `.create()` composable.

File: package.json (hunk 8 above)
Comment (ksafranski) [reply package.json]: Good call, I'll pull it in.

**Comment by levithomason** [comment]
Haha, oh man, this has been a revealing experience for PH tests...

**Comment by ksafranski** [comment]
I'm sure. Guess that means this PR is on the right track?

**Comment by ksafranski** [comment]
Got in some changes/fixes, additional tests and redid the documentation. Once we get sign-off on the PH testing I think we can merge and release a 2.x

**Comment by ksafranski** [comment]
I just added a `Proxy` around the `fixd` object on export, in conjunction with nesting `add`'d values in an frozen object under prop `$fixdVal`. This allows **any** type of value to be stored in a frozen state.

File: README.md [hunk 9]
```diff
@@ -12,92 +12,33 @@ Fixd is a JavaScript utility for creating better fixtures. Fixtures created with
...
 
-// Create a new instance of the fixture
-const fooOne = fixd.create('foo') // -> { foo: 'bar' }
+// Access fixture
+const foo = fixd.foo // -> { fizz: 'buzz', baz: { quz: 'baz' } }
 
-// Create a different instance with alterations to original object
-const fooTwo = fixd.create('foo', { foo: 'baz' }) // -> { foo: 'baz' }
+// Throws error
+foo.fizz = 'bazz' // -> ERROR: Cannot assign to read only property
```
Comment (dvdzkwsk) [inline README.md]: Since this is even suggested in the docs, can't we allow this as a one-time assignment?

```js
fixd.foo = {} // ok! same as `fixd.add`
fixd.foo = {} // nope! Already set.
```

I suggested the same to @levithomason since it would offer parity w/ the getter, making the API more symmetrical aside from the `create` for mutations.

File: README.md [hunk 10]
```diff
@@ -12,92 +12,33 @@ Fixd is a JavaScript utility for creating better fixtures. Fixtures created with
...
-
-const fizzFixture = fixd.create('fizz') // -> { "buzz": true }
-```
-
-On create, an optional (`Object`) argument can be passed to alter the original fixture:
-
-```javascript
-const fizzFixture = fixd.create('fizz', { zap: 'bang' }) // -> { "buzz": true, "zap": "bang" }
-```
+The above will not modify the `foo` object, instead will return a new, reference-free, frozen instance of the object.
```
Comment (dvdzkwsk) [inline README.md]: `[.] Instead, [it]`.

File: fixd.js [hunk 11]
```diff
@@ -1,14 +1,55 @@
...
+  add: (name, val) => {
+    if (fixd[name]) {
+      throw new Error(`Cannot add fixture to already reserved namespace '${name}'.`)
+    }
+    fixd[name] = fixd.freeze({ $fixdVal: val })
+  },
+  /**
+   * Returns a fixture and allows for (option) modification
+   * @param {String} name The namespace of the fixture to mutate
+   * @param {Function} modifier A function to modify the fixture
```
Comment (dvdzkwsk) [inline fixd.js:53]: `* @returns`

File: fixd.js [hunk 12]
```diff
@@ -1,14 +1,55 @@
-const fs = require('fs')
-const path = require('path')
-const fixd = require('./src/index')
+'use strict'
+const merge = require('lodash.merge')
```
Comment (dvdzkwsk) [inline fixd.js]: What about `clone` instead of `merge`? That way you can avoid your `Array.isArray` check and defer any polymorphism to lodash. Not a blocker, but an idea.

File: fixd.js [hunk 13]
```diff
@@ -1,14 +1,55 @@
...
+  create: (name, modifier) => {
+    if (typeof modifier !== 'function') {
+      throw new Error('Modifier must be a function')
+    }
+    return fixd.freeze(modifier(merge(Array.isArray(fixd[name].$fixdVal) ? [] : {}, fixd[name].$fixdVal)))
+  }
 }
 
-module.exports = fixd
+module.exports = new Proxy(fixd, {
```
Comment (dvdzkwsk) [inline fixd.js]: Proxy is great and all, but you may be limiting yourself from greater public adoption, especially from front-end codebases that run unit tests in older browsers. I personally love Proxy and wish I could use it more, but it's something to consider. Other option is to wait until somebody actually experiences an issue to do anything about it, which I think is fair just as well.

File: fixd.js [hunk 14]
```diff
@@ -1,14 +1,55 @@
...
+  get: function (target, name) {
+    if (typeof name === 'symbol') return target
```
Comment (dvdzkwsk) [inline fixd.js]: Why can't you do `target[name]` if the name is a symbol?

File: fixd.js (hunk 13 above)
Comment (ksafranski) [reply fixd.js]: Yeah, it was a consideration, however, I feel like this is the perfect use-case, and while it could limit adoption, I'd prefer lean forward than back. I'm happy to get this in for now (for our use-cases) and worry about it should we get an issue.

File: fixd.js (hunk 14 above)
Comment (ksafranski) [reply fixd.js]: The proxy will look for `target[name].$fixdVal` if you try to just get `fixd`. The `typeof` match on `symbol` allows it to return the entire `fixd` object

File: fixd.js (hunk 13 above)
Comment (dvdzkwsk) [reply fixd.js]: Felt the same way (leaning forward) earlier. I'm down.

File: fixd.js (hunk 12 above)
Comment (ksafranski) [reply fixd.js]: Will do

File: README.md (hunk 9 above)
Comment (ksafranski) [reply README.md]: Yeah, good call, now that I have the Proxy this will be cake.

Comment (davezuko) [inline README.md]: ^^ Also, gives you more reason to use proxies :).

File: fixd.js (hunk 11 above)
Comment (ksafranski) [reply fixd.js:53]: Done.

File: README.md (hunk 9 above)
Comment (ksafranski) [reply README.md]: Done.

File: README.md [hunk 15]
```diff
@@ -3,102 +3,56 @@
...
+fixd.foo // -> { fizz: 'buzz', baz: { quz: 'baz' } }
 
-// Create a different instance with alterations to original object
-const fooTwo = fixd.create('foo', { foo: 'baz' }) // -> { foo: 'baz' }
+// Throws error, prevents mutations
+foo.fizz = 'bazz' // -> ERROR: Cannot assign to read only property
 ```
 
-## Plugins
+_Note: While `Object.freeze` will prevent modifications, `use strict` must be applied to the document in order to throw error on attempted mutation._
```
Comment (levithomason) [inline README.md]: `in order to throw [an] error`

File: README.md [hunk 16]
```diff
@@ -3,102 +3,56 @@
...
 ```javascript
-const fooFixture = fixd.create('foo', { foo: 'bam' }) // -> { foo: 'bam' }
+fixd.create('foo', (foo) => {
+  foo.fizz = 'bazz'
+  return foo
+}) // -> { fizz: 'bazz', baz: { quz: 'baz' } }
 ```
 
-### Array
+The above will not modify the `foo` object, it will return a new, reference-free, frozen instance of the object based on the modification performed in the function.
```
Comment (levithomason) [inline README.md]: Suggestion: `in the function` => `in the callback`

File: fixd.js [hunk 17]
```diff
@@ -1,14 +1,61 @@
...
+  },
+  /**
+   * Retrieves a value from the fixd library
+   * @param {Object} target Target object to retrieve from
+   * @param {String} name Property name to retrieve
+   * @returns {*} The value of the property
+   */
+  get: (target, name) => {
+    if (typeof name === 'symbol') return target
+    if (['freeze', 'get', 'set', 'create'].indexOf(name) >= 0) return target[name]
```
Comment (levithomason) [inline fixd.js]: How about:`[].includes()`?

File: fixd.js [hunk 18]
```diff
@@ -1,14 +1,61 @@
...
-  throw new Error('Error loading built-in plugins', e.message)
+const fixd = {
+  /**
+   * Deeply freezes an object
+   * @param {*} obj The object to freeze
+   * @returns {*}
+   */
+  freeze: (obj) => {
+    Object.freeze(obj)
+    Object.getOwnPropertyNames(obj).forEach((prop) => {
```
Comment (levithomason) [inline fixd.js]: This method throws when called with `null` or `undefined` due to `getOwnPropertyNames`.
```js
fixd.foo = { bar: 'bar' }

fixd.create('foo', (foo) => {
  return null
})
// Object.getOwnPropertyNames(obj).forEach((prop) => {
//        ^
// TypeError: Cannot convert undefined or null to object

fixd.create('foo', (foo) => {
  return undefined
})
// Object.getOwnPropertyNames(obj).forEach((prop) => {
//        ^
// TypeError: Cannot convert undefined or null to object
```

Comment (ksafranski) [reply fixd.js]: Fixed by checking `obj`

**Comment by levithomason** [comment]
Suggest we add test cases for `create()`ing `null` and `undefined` values:

```js
fixd.create('foo', () => null)
// => TypeError: Cannot read property '$fixdVal' of undefined

fixd.create('foo', () => undefined)
// => TypeError: Cannot read property '$fixdVal' of undefined
```

**EDIT** Whoops, these are actually non-existent keys...

### create null/undefined
```js
fixd.foo = { foo: true }

fixd.create('foo', () => null)
// => TypeError: Cannot convert undefined or null to object

fixd.create('foo', () => undefined)
// => TypeError: Cannot convert undefined or null to object
```

**Comment by ksafranski** [comment]
^^^ in progress

**Comment by dvdzkwsk** [comment]
Returning `null` or `undefined` from `create` should probably throw an error explaining to the user that they've forgotten to return a value from their callback. I can't see a use case for having a `null` or `undefined` fixture. If they want to use either of those values, there's no use for fixd.

**Comment by levithomason** [comment]
Alright, I've hit battered this from every angle I can think of right now.  LMK when the latest update is pushed and I'll hop back over to PH and give it a whirl.

Nice work with the Proxy!

**Comment by levithomason** [comment]
I like the error for returning nil from `create()`.

Similarly, how about a helpful error when attempting to `create()` a non-existent namespace, as I did above?

File: fixd.js [hunk 19]
```diff
@@ -56,7 +57,9 @@ const fixd = {
     if (typeof modifier !== 'function') {
       throw new Error('Modifier must be a function')
     }
-    return fixd.freeze(modifier(clone(fixd[name].$fixdVal)))
+    const newObj = modifier(clone(fixd[name].$fixdVal))
+    if (newObj) return fixd.freeze(newObj)
+    throw new Error('Modifier should return a value')
```
Comment (levithomason) [inline fixd.js]: `if (newObj)` doesn't account for falsy values, which may be valid returns.  Suggest checking for nil (null/undefined) explicitly.

**Comment by levithomason** [comment]
Retrying IRL testing

**Comment by ksafranski** [comment]
Ok, another round of changes applied...

**Comment by levithomason** [comment]
yep, https://github.com/manosim/gitify 😄
